client-lite/src/ipc/rest_http_listener_conn.cpp

```diff
@@ -0,0 +1,139 @@
+#include "do_common.h"
+#include "rest_http_listener_conn.h"
+
+#include "do_http_defines.h"
+#include "do_version.h"
+
+using boost_tcp_t = boost::asio::ip::tcp;
+namespace msdod = microsoft::deliveryoptimization::details;
+
+HttpListenerConnection::HttpListenerConnection(boost::asio::io_service& ioService, std::shared_ptr<boost::asio::ip::tcp::socket> socket) :
+    _socket(std::move(socket)),
+    _io(ioService)
+{
+    _recvBuf.resize(2048);
+}
+
+HttpListenerConnection::~HttpListenerConnection()
+{
+    if (_socket->is_open())
+    {
+        std::stringstream ss;
+        ss << _socket->remote_endpoint();
+        DoLogDebug("Socket closing: was connected to %s", ss.str().c_str());
+        _socket->shutdown(boost_tcp_t::socket::shutdown_both);
+        _socket->close();
+    }
+}
+
+std::shared_ptr<HttpListenerConnection> HttpListenerConnection::Make(boost::asio::io_service& ioService,
+    std::shared_ptr<boost::asio::ip::tcp::socket> socket)
+{
+    return std::make_shared<HttpListenerConnection>(ioService, std::move(socket));
+}
+
+void HttpListenerConnection::Receive(http_listener_callback_t& callback)
+{
+    _socket->async_read_some(boost::asio::buffer(_recvBuf.data(), _recvBuf.size()),
+        [this, lifetime = shared_from_this(), &callback](const boost::system::error_code& ec, size_t cbRead)
+            {
+                _OnData(ec, cbRead, callback);
+            });
+}
+
+void HttpListenerConnection::Reply(unsigned int statusCode)
+{
+    Reply(statusCode, std::string{});
+}
+
+static const char* g_HttpStatusToString(unsigned int statusCode)
+{
+#define RETURN_HTTP_STATUSCODE_STR(code) \
+    case msdod::http_status_codes::code: return #code;
+
+    switch (statusCode)
+    {
+        RETURN_HTTP_STATUSCODE_STR(OK);
+        RETURN_HTTP_STATUSCODE_STR(BadRequest);
+        RETURN_HTTP_STATUSCODE_STR(NotFound);
+        RETURN_HTTP_STATUSCODE_STR(InternalError);
+        RETURN_HTTP_STATUSCODE_STR(ServiceUnavailable);
+        default: return "StatusDescription";
+    }
+    return "StatusDescription";
+}
+
+void HttpListenerConnection::Reply(unsigned int statusCode, const std::string& body)
+{
+    std::stringstream ss;
+    ss << "HTTP/1.1 " << statusCode << ' ' << g_HttpStatusToString(statusCode) << "\r\n";
+    if (!body.empty())
+    {
+        ss << "Content-Length: " << body.size() << "\r\n";
+    }
+    ss << "Server: Delivery-Optimization-Agent/" << microsoft::deliveryoptimization::util::details::SimpleVersion() << "\r\n";
+    ss << "\r\n";
```

Comment:
This crlf is used for the empty header correct?

Reply:
Yes, it signifies the end of the header section.